lib/Dialect/HW/HWOps.cpp

```diff
@@ -1785,6 +1785,107 @@ LogicalResult ArrayCreateOp::verify() {
   return success();
 }
 
+// Check whether an integer value is an offset from a base.
+static bool isOffset(Value base, Value index, uint64_t offset) {
+  if (auto constBase = base.getDefiningOp<hw::ConstantOp>()) {
+    if (auto constIndex = index.getDefiningOp<hw::ConstantOp>()) {
+      // If both values are a constant, check if index == base + offset.
+      // To account for overflow, the addition is performed with an extra bit
+      // and the offset is asserted to fit in the bit width of the base.
+      auto baseValue = constBase.getValue();
+      auto indexValue = constIndex.getValue();
+
+      unsigned bits = baseValue.getBitWidth();
+      assert(bits == indexValue.getBitWidth() && "mismatched widths");
+
+      if (bits < 64 && offset >= (1ull << bits))
+        return false;
+
+      APInt baseExt = baseValue.zextOrTrunc(bits + 1);
+      APInt indexExt = indexValue.zextOrTrunc(bits + 1);
+      return baseExt + offset == indexExt;
+    }
+  }
+  return false;
+}
+
+// Canonicalize a create of consecutive elements to a slice.
+static LogicalResult foldCreateToSlice(ArrayCreateOp op,
+                                       PatternRewriter &rewriter) {
+  // Do not canonicalize create of get into a slice.
+  auto arrayTy = hw::type_cast<ArrayType>(op.getType());
+  if (arrayTy.getSize() <= 1)
+    return failure();
+  auto elemTy = arrayTy.getElementType();
+
+  // Check if create arguments are consecutive elements of the same array.
+  // Attempt to break a create of gets into a sequence of consecutive intervals.
+  struct Chunk {
+    Value input;
+    Value index;
+    size_t size;
+  };
+  SmallVector<Chunk> chunks;
+  for (Value value : llvm::reverse(op.getInputs())) {
+    auto get = value.getDefiningOp<ArrayGetOp>();
+    if (!get)
+      return failure();
+
+    Value input = get.getInput();
+    Value index = get.getIndex();
+    if (!chunks.empty()) {
+      auto &c = *chunks.rbegin();
+      if (c.input == get.getInput() && isOffset(c.index, index, c.size)) {
+        c.size++;
+        continue;
+      }
+    }
+
+    chunks.push_back(Chunk{input, index, 1});
+  }
+
+  // If there is a single slice, eliminate the create.
+  if (chunks.size() == 1) {
+    Value input = chunks[0].input;
+    Value index = chunks[0].index;
+
+    // If the whole array is sliced, return it in full.
+    if (input.getType() == op.getType()) {
+      rewriter.replaceOp(op, input);
+      return success();
+    }
```

Comment:
nit: We have a folder to fold identity array slice (https://github.com/llvm/circt/blob/680ef379c81a7c8a7e64e95833e7981d71dfc71f/lib/Dialect/HW/HWOps.cpp#L1813-L1818) so you can simplify this if you want:

```c++
rewriter.replaceOp(op, rewriter.createOrFold<ArraySliceOp>(op.getLoc(), arrayTy, input, index));
```

Reply:
Awesome, thanks!